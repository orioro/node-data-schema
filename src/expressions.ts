import {
  EvaluationContext,
  Expression,
  $$VALUE,
  evaluate
} from '@orioro/expression'

const defaultGetType = () => undefined

export const schemaTypeExpression = (getType = defaultGetType) => (
  context:EvaluationContext,
  valueExp:Expression = $$VALUE
) => {
  const value = evaluate(context, valueExp)
  let type = getType(value)

  if (typeof type === 'string') {
    return type
  } else {

    type = typeof value

    if (type === 'object') {
      if (Array.isArray(type)) {
        return 'list'
      } else {
        return 'map'
      }

    } else {
      return type
    }
  }
}
