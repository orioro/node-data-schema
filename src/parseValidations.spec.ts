import {
  allowValues,
  prohibitValues,
  parallelCases,
  validate
} from '@orioro/validate'

import { schemaTypeExpression } from './expressions'

import {
  ALL_EXPRESSIONS
} from '@orioro/expression'

import {
  parseValidations,
  mapValidationResolver,
  listValidationResolver,
  defaultValidationResolver,
  stringValidationResolver,
} from './parseValidations'

const dump = value => console.log(JSON.stringify(value, null, '  '))

const REQUIRED_ERROR = {
  code: 'REQUIRED_ERROR',
  message: 'This value is required'
}

const TYPE_ERROR = {
  code: 'TYPE_ERROR',
  message: 'Invalid type'
}

const interpreters = {
  ...ALL_EXPRESSIONS,
  $schemaType: schemaTypeExpression(),
}

describe('parseValidations(schema, context) - required / optional', () => {
  test('required', () => {

    const schema = {
      type: 'string',
      required: true,
      errors: {
        required: REQUIRED_ERROR,
        type: TYPE_ERROR
      }
    }

    const validations = parseValidations(schema, undefined, {
      resolvers: [
        mapValidationResolver(),
        defaultValidationResolver()
      ]
    })

    expect(validations).toEqual([
      {
        path: '',
        validation: prohibitValues(
          [null, undefined],
          REQUIRED_ERROR,
          [
            '$if',
            ['$notEq', 'string', ['$schemaType']],
            TYPE_ERROR,
            null
          ]
        )
      }
    ])

    const expectations = [
      [null, [REQUIRED_ERROR]],
      [undefined, [REQUIRED_ERROR]],
      [9, [TYPE_ERROR]],
      ['some text', null]
    ]

    expectations.forEach(([input, expected]) => {
      expect(validate(validations[0].validation, input, { interpreters }))
        .toEqual(expected)
    })
  })

  test('optional', () => {
    const schema = {
      type: 'string',
      errors: {
        type: TYPE_ERROR
      }
    }

    const validations = parseValidations(schema, undefined, {
      resolvers: [
        mapValidationResolver(),
        defaultValidationResolver()
      ]
    })

    expect(validations).toMatchObject([
      {
        path: '',
        validation: allowValues(
          [null, undefined],
          [
            '$if',
            ['$notEq', 'string', ['$schemaType']],
            { code: 'TYPE_ERROR' },
            null
          ]
        )
      }
    ])

    const expectations = [
      [null, null],
      [undefined, null],
      [9, [TYPE_ERROR]],
      ['some text', null]
    ]

    expectations.forEach(([input, expected]) => {
      expect(validate(validations[0].validation, input, { interpreters }))
        .toEqual(expected)
    })
  })

})

test('string validations', () => {
  const MIN_LENGTH_ERROR = {
    code: 'MIN_LENGTH_ERROR',
    message: 'Text must have at least 1 char'
  }

  const MAX_LENGTH_ERROR = {
    code: 'MAX_LENGTH_ERROR',
    message: 'Text must have at most 10 chars'
  }

  const PATTERN_ERROR = {
    code: 'PATTERN_ERROR',
    message: 'Text must match pattern'
  }

  const schema = {
    type: 'string',
    required: true,
    minLength: 5,
    maxLength: 10,
    pattern: ['^a.+z$', 'i'],
    errors: {
      required: REQUIRED_ERROR,
      type: TYPE_ERROR,
      minLength: MIN_LENGTH_ERROR,
      maxLength: MAX_LENGTH_ERROR,
      pattern: PATTERN_ERROR
    }
  }

  const validations = parseValidations(schema, 'Some text', {
    resolvers: [
      stringValidationResolver(),
      defaultValidationResolver()
    ]
  })

  expect(validations).toEqual([
    {
      path: '',
      validation: prohibitValues(
        [null, undefined],
        REQUIRED_ERROR,
        [
          '$if',
          ['$notEq', 'string', ['$schemaType']],
          TYPE_ERROR,
          parallelCases([
            [
              ['$gte', 5, ['$stringLength']],
              MIN_LENGTH_ERROR
            ],
            [
              ['$lte', 10, ['$stringLength']],
              MAX_LENGTH_ERROR
            ],
            [
              ['$stringTest', ['^a.+z$', 'i']],
              PATTERN_ERROR
            ]
          ])
        ]
      )
    }
  ])

  const expectations = [
    [null, [REQUIRED_ERROR]],
    [undefined, [REQUIRED_ERROR]],
    [9, [TYPE_ERROR]],
    ['a', [MIN_LENGTH_ERROR, PATTERN_ERROR]],
    ['abcdefghijklmnopqrstuv', [MAX_LENGTH_ERROR, PATTERN_ERROR]],
    ['bcdefghi', [PATTERN_ERROR]],
    ['abcdez', null],
    ['AbcdeZ', null],
  ]

  expectations.forEach(([input, expected]) => {
    expect(validate(validations[0].validation, input, { interpreters }))
      .toEqual(expected)
  })
})

describe('list validations', () => {
  const MIN_LENGTH_ERROR = {
    code: 'MIN_LENGTH_ERROR'
  }

  const MAX_LENGTH_ERROR = {
    code: 'MAX_LENGTH_ERROR'
  }

  const schema = {
    type: 'list',
    itemSchema: {
      type: 'string',
      required: true,
      minLength: 5,
      maxLength: 10,
      errors: {
        required: REQUIRED_ERROR,
        type: TYPE_ERROR,
        minLength: MIN_LENGTH_ERROR,
        maxLength: MAX_LENGTH_ERROR
      }
    }
  }

  test('', () => {
    const validations = parseValidations(schema, ['123', '12345', '1234567', '123456789012345'], {
      resolvers: [
        listValidationResolver(),
        stringValidationResolver(),
        defaultValidationResolver()
      ]
    })

    expect(validations).toMatchSnapshot()
  })
})
