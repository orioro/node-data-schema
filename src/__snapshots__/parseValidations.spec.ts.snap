// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list validations  1`] = `
Array [
  Object {
    "path": "",
    "validation": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      null,
      Array [
        "$if",
        Array [
          "$notEq",
          "list",
          Array [
            "$schemaType",
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`list\`",
        },
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
      ],
    ],
  },
  Object {
    "path": "0",
    "validation": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$notEq",
          "string",
          Array [
            "$schemaType",
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
      ],
    ],
  },
  Object {
    "path": "1",
    "validation": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$notEq",
          "string",
          Array [
            "$schemaType",
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
      ],
    ],
  },
  Object {
    "path": "2",
    "validation": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$notEq",
          "string",
          Array [
            "$schemaType",
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
      ],
    ],
  },
  Object {
    "path": "3",
    "validation": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$notEq",
          "string",
          Array [
            "$schemaType",
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
      ],
    ],
  },
]
`;
